src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -729,7 +756,7 @@ private void InheritNullableStateOfMember(int targetContainerSlot, int valueCont
 
             TypeSymbolWithAnnotations fieldOrPropertyType = member.GetTypeOrReturnType();
 
-            if (fieldOrPropertyType.IsReferenceType)
+            if (fieldOrPropertyType.IsReferenceType || fieldOrPropertyType.IsNullableType())
```

Comment:
> || fieldOrPropertyType.IsNullableType() [](start = 51, length = 40)

This doesn't feel right. It looks like this special case is needed because simply discounts all "special" types:
```
        public static bool IsTrackableStructType(TypeSymbol type)
        {
            if ((object)type == null) return false;
            var nts = type.OriginalDefinition as NamedTypeSymbol;
            if ((object)nts == null) return false;
            return nts.IsStructType() && nts.SpecialType == SpecialType.None && !nts.KnownCircularStruct;
        }
```

I assume the ```nts.SpecialType == SpecialType.None``` is just an optimization that is supposed to target circular "special" structures (System.Int32, etc.). We shouldn't assume that all special structures like that and instead optimize for specific kinds. I see other places in NullableWalker where EmptyStructTypeCache.IsTrackableStructType is used and those places might be negatively affected as well.

Also, I am not sure if we are tracking this issue already, but it looks like we have a problem with tracking properties on structures, this might have impact on analysis around nullable value types:
```
            CSharpCompilation c = CreateCompilation(new[] { @"
class C
{
    static void Main()
    {
    }

    void Test1(S1 x1)
    {
        x1.F1.ToString(); // 1
        x1.F1.ToString(); // 2
        x1.F1 = """";
        x1.F1.ToString(); // 3
    }
}

#pragma warning disable CS0649
struct S1
{
public string? F1
{
    get
    {
        throw null;
    }
    set{}
}
public string? f1;
}
" }, options: WithNonNullTypesTrue());
#if EXPECTED
            c.VerifyDiagnostics(
                // (10,9): warning CS8602: Possible dereference of a null reference.
                //         x1.F1.ToString(); // 1
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x1.F1").WithLocation(10, 9)
                );
#else
    // ACTUAL
            c.VerifyDiagnostics(
                // (10,9): warning CS8602: Possible dereference of a null reference.
                //         x1.F1.ToString(); // 1
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x1.F1").WithLocation(10, 9),
                // (11,9): warning CS8602: Possible dereference of a null reference.
                //         x1.F1.ToString(); // 2
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x1.F1").WithLocation(11, 9),
                // (13,9): warning CS8602: Possible dereference of a null reference.
                //         x1.F1.ToString(); // 3
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x1.F1").WithLocation(13, 9)
                );
#endif

```

Reply:
We are only tracking auto-properties for structs currently (see [https://github.com/dotnet/roslyn/issues/29619](https://github.com/dotnet/roslyn/issues/29619)) which is the reason for the special case here.

---
In reply to: [239921344](https://github.com/dotnet/roslyn/pull/31499#discussion_r239921344) [](ancestors = 239921344)

Reply:
> We are only tracking auto-properties for structs currently (see [https://github.com/dotnet/roslyn/issues/29619](https://github.com/dotnet/roslyn/issues/29619) which is the reason for the special case here.

I think this particular special case should be explicitly mentioned in the issue and probably explicitly mentioned in a comment in source code 

---
In reply to: [239996177](https://github.com/dotnet/roslyn/pull/31499#discussion_r239996177) [](ancestors = 239996177,239921344)